### PR TITLE
Make DeclRefBase a Val, and DeclRef<T> a helper class.

### PR DESCRIPTION
**Pull request by csyonghe**
The intention of this change is to make DeclRef itself a syntax node so it can participate in deduplication, and can later be extended to have different subclasses to represent different cases of a DeclRef.

The main body of work in this change is that since DeclRef is a syntax node itself, we need to take extra care to make sure dynamic casts around `DeclRef` are still having its original meaning, that is casting the Decl being pointed to instead of the DeclRef object itself.

### CODE REVIEW

File: source/slang/slang-ast-base.h [hunk 1]
```diff
@@ -405,20 +489,22 @@ class Decl : public DeclBase
 
     ContainerDecl* parentDecl = nullptr;
 
+    DeclRefBase* defaultDeclRef = nullptr;
```
Comment (tangent-vector) [inline source/slang/slang-ast-base.h:492]: Could use a comment indicating what this is/means.

File: source/slang/slang-ast-base.h [hunk 2]
```diff
@@ -446,5 +532,115 @@ class Stmt : public ModifiableSyntaxNode
...
+    DeclRefBase* declRef = nullptr;
+    if (decl)
+    {
+        SLANG_ASSERT(decl->defaultDeclRef);
+        declRef = decl->defaultDeclRef;
+    }
+    else
+    {
+        declRef = nullptr;
+    }
```
Comment (tangent-vector) [inline source/slang/slang-ast-base.h:556]: This side is a no-op, given that `declRef` is already initialized to null above, right?

File: source/slang/slang-ast-base.h [hunk 3]
```diff
@@ -446,5 +532,115 @@ class Stmt : public ModifiableSyntaxNode
...
+HashCode DeclRef<T>::getHashCode() const
+{
+    if (declRefBase) return declRefBase->getHashCode();
+    return 0;
+}
+
+template<typename T>
+Type* DeclRef<T>::substitute(ASTBuilder* astBuilder, Type* type) const
+{
+    SLANG_ASSERT(declRefBase);
```
Comment (tangent-vector) [inline source/slang/slang-ast-base.h:611]: I suppose that a null `DeclRef` could act as a no-op set of substitutions, but making it an invariant that the decl ref is non-null (as this code does) certainly seems safer, if there isn't existing code that violates that assumption.

File: source/slang/slang-ast-builder.h [hunk 4]
```diff
@@ -254,34 +252,39 @@ class ASTBuilder : public RefObject
...
     {
-        return DeclRef<T>(this, decl, subst);
+        // We never create an actual DeclRefBase node to point to a null decl.
+        if (!decl)
+            return DeclRef<T>();
+
+        // If we don't have substitutions, use the default decl ref if it is created.
+        if (!subst)
+        {
+            auto& defaultDeclRef = static_cast<Decl*>(decl)->defaultDeclRef;
```
Comment (tangent-vector) [inline source/slang/slang-ast-builder.h:266]: Why is a reference used here?

**Review by tangent-vector — DISMISSED** [review]
There's a lot of churn here, but most of it is straightforward.

There's a couple places where there seem to be larger code changes, and maybe there's subtle stuff there I'm missing, but in most cases it seems to be adapting code that was mutating `DeclRef`s to instead construct fresh ones.

File: source/slang/slang-ast-base.h (hunk 2 above)
Comment (csyonghe) [reply source/slang/slang-ast-base.h]: Fixed.

File: source/slang/slang-ast-base.h (hunk 1 above)
Comment (csyonghe) [reply source/slang/slang-ast-base.h:497]: Commented.

File: source/slang/slang-ast-builder.h (hunk 4 above)
Comment (csyonghe) [reply source/slang/slang-ast-builder.h]: Fixed.

File: source/slang/slang-ast-base.h (hunk 3 above)
Comment (csyonghe) [reply source/slang/slang-ast-base.h]: Turned it into a no-op case.

**Pull request merged**